src/ansi-c/c_typecheck_code.cpp

```diff
@@ -842,35 +844,147 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
+    error().source_location = target.source_location();
+    error() << "void-typed expressions not allowed as assigns clause targets"
+            << eom;
+  }
+
+  if(!target.get_bool(ID_C_lvalue) && target.id() != ID_pointer_object)
+  {
+    must_throw = true;
+    error().source_location = target.source_location();
+    error() << "assigns clause target must be an lvalue or " CPROVER_PREFIX
+               "POINTER_OBJECT expression"
+            << eom;
+  }
+
+  // Remark: an expression can be an lvalue and still contain side effects
+  // or ternary expressions. We detect them in a second step.
+  if(has_subexpr(target, ID_side_effect))
+  {
+    must_throw = true;
+    error().source_location = target.source_location();
+    error() << "side-effects not allowed in assigns clause targets" << eom;
+  }
+
+  if(has_subexpr(target, ID_if))
+  {
+    must_throw = true;
+    error().source_location = target.source_location();
+    error() << "ternary expressions not allowed in assigns "
+               "clause targets"
+            << eom;
+  }
+
+  if(must_throw)
+    throw 0;
+}
+
 void c_typecheck_baset::typecheck_spec_assigns(exprt::operandst &targets)
 {
+  exprt::operandst tmp;
   for(auto &target : targets)
   {
-    typecheck_expr(target);
-
-    if(target.type().id() == ID_empty)
+    // conditional target
+    if(can_cast_expr<conditional_assigns_target_exprt>(target))
     {
-      error().source_location = target.source_location();
-      error() << "void-typed targets not permitted in assigns clause" << eom;
-      throw 0;
-    }
-    else if(target.id() == ID_pointer_object)
-    {
-      // skip
-    }
-    else if(!target.get_bool(ID_C_lvalue))
-    {
-      error().source_location = target.source_location();
-      error() << "non-lvalue target in assigns clause" << eom;
-      throw 0;
+      auto &conditional_target = to_conditional_assigns_target_expr(target);
+
+      // typecheck condition
+      auto &condition = conditional_target.condition();
+      typecheck_spec_assigns_condition(condition);
+
+      // typecheck target(s)
+      if(conditional_target.target().id() == ID_expression_list)
+      {
+        // distribute condition over all actual targets
+        for(auto &actual_target : conditional_target.target().operands())
+        {
+          typecheck_spec_assigns_target(actual_target);
+          conditional_assigns_target_exprt result{condition, actual_target};
+          tmp.push_back(std::move(result));
+        }
+      }
+      else
+      {
+        typecheck_spec_assigns_target(conditional_target.target());
+        tmp.push_back(std::move(conditional_target));
+      }
     }
-    else if(has_subexpr(target, ID_side_effect))
+    else // unconditional target
     {
-      error().source_location = target.source_location();
-      error() << "assigns clause is not side-effect free" << eom;
-      throw 0;
+      typecheck_spec_assigns_target(target);
+      tmp.push_back(std::move(target));
```

Comment:
Would it be of any value to make up a conditional target with condition `true_exprt{}` here?